src/number_theory/modular_forms/modular_forms.lean

```diff
@@ -0,0 +1,391 @@
+  rw this,
+  apply slash_k_mul_one,
+end
+
+lemma smul_slash_k (k : ℤ) (A : GL(2, ℝ)⁺) (f : ℍ → ℂ) (c : ℂ) : (c • f) ∣[k] A = c • (f ∣[k] A) :=
+begin
+  ext1,
+  simp_rw slash_k,
+  simp only [slash_k, algebra.id.smul_eq_mul, matrix.general_linear_group.coe_det_apply,
+    pi.smul_apply, subtype.val_eq_coe, coe_coe],
+  ring,
+end
+
+lemma slash_k_mul (k1 k2 : ℤ) (A : GL(2, ℝ)⁺) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1+k2] A = (((↑ₘ A).det) : ℝ) • (f ∣[k1] A) * (g ∣[k2] A) :=
+begin
+  ext1,
+  simp only [slash_k, matrix.general_linear_group.coe_det_apply, subtype.val_eq_coe, ←mul_assoc],
+  have h1 : ((((↑ₘ A).det) : ℝ)^(k1 + k2 - 1) : ℂ) =
+  (((↑ₘ A).det) : ℝ) * (((↑ₘ A).det) : ℝ)^(k1 - 1) * (((↑ₘ A).det) : ℝ)^(k2 - 1),
+  by {simp only [mul_assoc, matrix.general_linear_group.coe_det_apply, subtype.val_eq_coe, coe_coe],
+  rw [←zpow_add₀, ←zpow_one_add₀],
+  ring_exp,
+  all_goals { norm_cast, apply (matrix.GL_pos.det_ne_zero A), }, },
+  have h22 : (upper_half_plane.denom A x)^(-(k1+k2)) = (upper_half_plane.denom A x)^(-k1) *
+    (upper_half_plane.denom A x)^(-k2),
+  by { rw [int.neg_add, zpow_add₀], exact upper_half_plane.denom_ne_zero A x, },
+  rw [h1,h22],
+  simp only [upper_half_plane.denom, pi.mul_apply, coe_coe, zpow_neg, algebra.smul_mul_assoc,
+    pi.smul_apply, real_smul],
+  ring,
+end
+
+lemma slash_k_mul_SL2 (k1 k2 : ℤ) (A : SL(2,ℤ)) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1 + k2] A = (f ∣[k1] A) * (g ∣[k2] A) :=
+begin
+  have : (((↑ₘ(A : GL(2,ℝ)⁺)).det) : ℝ) = 1,
+  { simp only [coe_coe,matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.det_coe], },
+  simp_rw [slash_k_mul, this, one_smul],
+end
+
+/--The space of functions that are modular-/
+def weakly_modular_submodule (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+  {carrier := { f : (ℍ → ℂ) | ∀ (γ : Γ), (f ∣[k] (γ : GL(2, ℝ)⁺)) = f },
+  zero_mem' := by { simp only [slash_k, set.mem_set_of_eq, pi.zero_apply, zero_mul, forall_const],
+    refl, },
+  add_mem' := by { intros f g hf hg γ,
+    rw [slash_k_add k γ f g, hf γ, hg γ], },
+  smul_mem' := by { intros c f hf γ,
+    have : (c • f) ∣[k] γ = c • (f ∣[k] γ), by {apply smul_slash_k},
+    rw (hf γ) at this,
+    apply this,} }
+
+lemma wmodular_mem (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) :
+  f ∈ (weakly_modular_submodule k Γ) ↔ ∀ (γ : Γ), (f ∣[k] (γ : GL(2, ℝ)⁺)) = f := iff.rfl
+
+lemma slash_k_mul_subgroup (k1 k2 : ℤ) (Γ : subgroup SL(2,ℤ)) (A : Γ) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1+k2] A = (f ∣[k1] A) * (g ∣[k2] A) :=
+begin
+  have : (((↑ₘ(A : GL(2,ℝ)⁺)).det) : ℝ) = 1,
+  by { simp only [coe_coe,matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.det_coe], },
+  simp_rw [slash_k_mul, this, one_smul],
+end
+
+/--A function `f : ℍ → ℂ` is modular, of  weight `k ∈ ℤ` and level `Γ`, if for every matrix in
+ `γ ∈ Γ` we have `f(γ • z)= (c*z+d)^k f(z)` where `γ= ![![a, b], ![c, d]]`,
+ and it acts on `ℍ` via Moebius trainsformations. -/
+@[simp]
```

Comment:
I am not sure this should be `simp`. The RHS is mathematically the definition of the LHS, right? We don't want `simp` to unfold definitions, this discourage writing a good API.

Reply:
Aha ok, thanks!